### DISCUSSION TITLE
Custom colors works if "hard coded", but not when iterating

### DISCUSSION THREAD

**Discussion by DavidCibin** [tailwindlabs/tailwindcss#12085] [Help]
```
/** @type {import('tailwindcss').Config} */
import plugin from "tailwindcss/plugin"

module.exports = {
    content: [
        "./src/pages/**/*.{js,ts,jsx,tsx,mdx}",
        "./src/components/**/*.{js,ts,jsx,tsx,mdx}",
        "./src/app/**/*.{js,ts,jsx,tsx,mdx}",
    ],
    theme: {
        extend: {
            colors: {
                "test-blue": "#1B75BB",
                "test-green": "#19A137",
                "test-yellow": "#EE9D08",
                "test-purple": "#A146BA",
                "test-red": "#C63530",
                "test-light-blue": "#4CA3E6",
                "test-light-green": "#38C857",
                "test-light-yellow": "#C6B28B",
                "test-light-purple": "#C388D3",
                "test-light-red": "#DB6F6B",
            },
        },
    },
    plugins: [
        plugin(function ({ addVariant }) {
            addVariant("david", ".david &")
        }),
    ],
}
```

At first, I added the colors array and used map() to iterate each color. But it didn't work, so tried to clear cache/reload and still nothing. Inspecting you can see the class were added to each element as it should, but no background color at all

```
const colors = [ "blue", "green", "yellow", "purple", "red"]

<div className="flex gap-2.5 ">
    {colors.map((color: string) => {
        return (
            <div key={color} className="h-12 w-12 bg-test-${color} david:bg-test-light-${color}`}></div>
        )
    })}
</div>
```

Then I added a div for each color, then everything works after that, even after removing the div's below, now the mapped color above works. SO what's going on here, it's like it had to be forced at first to the color be added to the css

```
<div className="flex gap-2.5 ">
    <div className="h-12 w-12 bg-test-blue david:bg-test-light-blue"></div>
    <div className="h-12 w-12 bg-test-green david:bg-test-light-green"></div>
    <div className="h-12 w-12 bg-test-yellow david:bg-test-light-yellow"></div>
    <div className="h-12 w-12 bg-test-purple david:bg-test-light-purple"></div>
    <div className="h-12 w-12 bg-test-red david:bg-test-light-red"></div>
    <div className="h-12 w-12 bg-test-violet david:bg-test-light-violet"></div>
</div>
```

Also, same issue happens if I toggle/add the 'david' mode. only worked after adding each div separately.

**Comment by wongjn** [comment]
As per [the documentation](https://tailwindcss.com/docs/content-configuration#dynamic-class-names):

> The most important implication of how Tailwind extracts class names is that it will only find classes that exist as _complete unbroken strings_ in your source files.
>
> If you use string interpolation or concatenate partial class names together, Tailwind will not find them and therefore will not generate the corresponding CSS:
>
> **Don’t construct class names dynamically**
>
> ```vue
> <div class="text-{{ error ? 'red' : 'green' }}-600"></div>
> ```
> In the example above, the strings `text-red-600` and `text-green-600` do not exist, so Tailwind will not generate those classes.
> Instead, make sure any class names you’re using exist in full:
>
> **Always use complete class names**
>
> ```vue
> <div class="{{ error ? 'text-red-600' : 'text-green-600' }}"></div>
> ```

**Reply by DavidCibin** [reply]
Thanks, as always: Read the damn documentation lol

**Comment by alex-krasikau** [comment]
Tailwind works by parsing your project files as plain text. It is a build tool and only has access to what is actually written in the files. That means that in order to add CSS declarations for the classes you use, Tailwind needs to see them at least once as a full class.

A common practice is to use full class names instead of constructing them using string templates:
```js
// Tailwind registers usage from strings in this array
const colors = [ "bg-test-blue", "bg-test-green", "bg-test-yellow", "bg-test-purple", "bg-test-red"]

<div className="flex gap-2.5 ">
    {colors.map((color: string) => {
        return (
            <div key={color} className={`h-12 w-12 ${color}`}></div>
        )
    })}
</div>
```

Alternatively you can add this options in a comment. Tailwind will see the classes and will add them to the resulting CSS file:
```js
// bg-test-blue, bg-test-green ...
// david:bg-test-light-blue david:bg-test-light-green ...
const colors = [ "blue", "green", "yellow", "purple", "red"]

<div className="flex gap-2.5 ">
    {colors.map((color: string) => {
        return (
            <div key={color} className="h-12 w-12 bg-test-${color} david:bg-test-light-${color}`}></div>
        )
    })}
</div>
```

**Reply by DavidCibin** [reply]
Thanks for the hint, adding it as comments could be helpful down the road

**Reply by alex-krasikau** [reply]
By the way, there are also good cases for using inline styles and custom properties. Maybe that would be helpful:
```js
const colors = [
  "#1B75BB",
  "#19A137",
  "#EE9D08",
  "#A146BA",
  "#C63530",
]

<div className="flex gap-2.5 ">
    {colors.map((color: string) => {
        return (
            <div key={color} style={{"--component-color": color}} className="h-12 w-12 bg-[--component-color]"></div>
        )
    })}
</div>
```
